File: Algorithms/golang-sorting/sort/insertionsort/insertionsort.go
package insertionsort

// Insertion sort is similar to how people sort their poker.
// Find a smaller one insert it at left part and shift
// Larger elements to the right side.
// Here used shift elements instead of Swap elements
// Which is more efficient than swap.
// Complexity is O(n2)
// If the array is sorted, it's O(n)
func InsertionSort(nums []int) {
	for i := 1; i < len(nums); i++ {
		j := i-1
		temp := nums[i]
		for j >=0 && nums[j] > temp {
			nums[j+1] = nums[j]
			j--
		}
		nums[j+1] = temp
	}
}

func InsertionSortWithIndexes(nums []int, left, right int) {
	for i := left+1; i < right; i++ {
		j := i-1
		temp := nums[i]
		for j >=0 && nums[j] > temp {
			nums[j+1] = nums[j]
			j--
		}
		nums[j+1] = temp
	}
}

File: Algorithms/Sorting/src/sorters/InsertionSort.java
package sorters;

import utility.Sortable;
import utility.Utility;

public class InsertionSort implements Sortable {
    private int[] nums;

    public InsertionSort(int[] arr){
        nums = arr;
    }
    public InsertionSort(){}

    @Override
    public int[] sort(){
        for(int i=1; i<nums.length; i++){
            int temp = nums[i];
            int j=i-1;
            while(j >=0 && nums[j] > temp){
                nums[j+1] = nums[j];
                j--;
            }
            nums[j+1] = temp;
        }
        return nums;
    }

    @Override
    public void setArray(int[] nums) {
        this.nums = nums;
    }

    static void sort(int[] array, int left, int right){
        for(int i=left+1; i<right+1; i++){
            int temp = array[i];
            int j=i-1;
            while(j >=left && array[j] > temp){
                array[j+1] = array[j];
                j--;
            }
            array[j+1] = temp;
        }
    }
}
